lti_consumer/plugin/views.py

```diff
@@ -420,3 +424,87 @@ def scores(self, request, *args, **kwargs):  # pylint: disable=unused-argument
             status=status.HTTP_201_CREATED,
             headers=headers
         )
+
+
+class LtiNrpsContextMembershipViewSet(viewsets.ReadOnlyModelViewSet):
+    """
+    LTI NRPS Context Membership Service endpoint.
+
+    See full documentation at:
+    http://imsglobal.org/spec/lti-nrps/v2p0
+    """
+
+    # Custom permission classes for LTI APIs
+    authentication_classes = [Lti1p3ApiAuthentication]
+    permission_classes = [LtiNrpsContextMembershipsPermissions]
+
+    # Renderer classes to accept LTI NRPS content types
+    renderer_classes = [
+        MembershipResultRenderer,
+    ]
+
+    def preprocess(self, data):
+        """
+        Preprocess the output of `get_membership` method. It now does followings -
+            - Append external ids to the user
+        """
+        processed_data = []
+
+        # batch get or create external ids for all users
+        user_ids = data.keys()
+        users = get_user_model().objects.prefetch_related('profile').filter(id__in=user_ids)
+
+        # get external ids
+        external_ids = compat.batch_get_or_create_externalids(users)
+
+        # import pdb; pdb.set_trace()
+        for userid, user_info in data.items():
+
+            # append external ids to user dictonary
+            user_info['external_id'] = external_ids[userid].external_user_id
+
+            processed_data.append(user_info)
+
+        return processed_data
```

Comment:
Simplify function, rename it to a proper name and add a better description. 
My suggestion also avoids copying the dictionary (a potentially large dataset) - the current code doubles the execution time compared to it.

Suggestion: 
```suggestion
    def attach_external_user_ids(self, data):
        """
        Process the output of `get_membership` method and append external ids to each
        user in the object.
        """
        # batch get or create external ids for all users
        user_ids = data.keys()
        users = get_user_model().objects.filter(id__in=user_ids)
        external_ids = compat.batch_get_or_create_externalids(users)
       
        # Attaches external id information to the original dict
        for userid in data:
            data[userid]['external_id'] = external_ids[userid].external_user_id
```

Reply:
Done.